### PR TITLE
Add command line option to get function pointer values from the value set abstract objects

### PR DESCRIPTION
**Pull request by NlightNFotis**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-analyzer/show_function_pointer_values.h
Comment (hannes-steffenhagen-diffblue) [inline src/goto-analyzer/show_function_pointer_values.h]: Newline at end of file

File: regression/goto-analyzer-get-fp-values/Makefile
Comment (hannes-steffenhagen-diffblue) [inline regression/goto-analyzer-get-fp-values/Makefile]: newline at end of file

File: regression/goto-analyzer-get-fp-values/CMakeLists.txt
Comment (hannes-steffenhagen-diffblue) [inline regression/goto-analyzer-get-fp-values/CMakeLists.txt]: newline at end of file

File: regression/goto-analyzer-get-fp-values/basic-usage-test/test.desc
Comment (hannes-steffenhagen-diffblue) [inline regression/goto-analyzer-get-fp-values/basic-usage-test/test.desc]: newline at end of file

File: regression/goto-analyzer-get-fp-values/chain.sh
Comment (hannes-steffenhagen-diffblue) [inline regression/goto-analyzer-get-fp-values/chain.sh]: newline at end of file

File: src/goto-analyzer/show_function_pointer_values.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/goto-analyzer/show_function_pointer_values.cpp]: newline at the end of file

File: src/goto-analyzer/show_function_pointer_values.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/goto-analyzer/show_function_pointer_values.cpp]: Missing module header, wrong copyright notice

File: src/goto-analyzer/show_function_pointer_values.h
Comment (hannes-steffenhagen-diffblue) [inline src/goto-analyzer/show_function_pointer_values.h]: Missing module header, wrong copyright notice

File: src/analyses/variable-sensitivity/variable_sensitivity_domain.h
Comment (thk123) [inline src/analyses/variable-sensitivity/variable_sensitivity_domain.h:117]: :question: (will check commit history) -> why is this necessary?

File: src/goto-analyzer/goto_analyzer_parse_options.cpp
Comment (thk123) [inline src/goto-analyzer/goto_analyzer_parse_options.cpp:274]: Some why comments here would be good (in particular, why verify and general analysis need to be set to true)

File: src/goto-analyzer/get_function_pointer_values.cpp
Comment (thk123) [inline src/goto-analyzer/get_function_pointer_values.cpp:161]: I would strongly like (perhaps in an immediate follow up PR) this function to be greatly simplified: having 6 levels of nesting and a nearly 100 line function are both code smells. At a minimum, each loop could be pulled out into a function that deals with one iteration

File: src/goto-analyzer/get_function_pointer_values.cpp
Comment (thk123) [inline src/goto-analyzer/get_function_pointer_values.cpp]: Use early return here since you only care about is_assign

File: src/goto-analyzer/get_function_pointer_values.cpp
Comment (thk123) [inline src/goto-analyzer/get_function_pointer_values.cpp:169]: This kind of data processing is where make_range comes into its own, and I'd strongly encourage taking some time to convert this to be a sequence of transformations

File: src/goto-analyzer/get_function_pointer_values.cpp
Comment (thk123) [inline src/goto-analyzer/get_function_pointer_values.cpp]: Again can use early return here

Comment (thk123) [inline src/goto-analyzer/get_function_pointer_values.cpp]: This whole block represents a function that can be easily pulled out that goes from identifier to boolean of whether it is a relevant assignment

File: src/goto-analyzer/get_function_pointer_values.cpp
Comment (thk123) [inline src/goto-analyzer/get_function_pointer_values.cpp]: :question: Doesn't feel like a precondition - perhaps an invariant?

File: src/goto-analyzer/get_function_pointer_values.cpp
Comment (thk123) [inline src/goto-analyzer/get_function_pointer_values.cpp]: :no_entry_sign: Please remove direct `std::cout` since breaks json/xml UIs (if really useful, `#ifdef DEBUG` - but in this instance, seems not generally useful, is just "trace" style info)

File: src/goto-analyzer/get_function_pointer_values.cpp
Comment (thk123) [inline src/goto-analyzer/get_function_pointer_values.cpp]: This can be pulled out into a function that goes from `lhs` to `values`

File: src/goto-analyzer/get_function_pointer_values.cpp
Comment (thk123) [inline src/goto-analyzer/get_function_pointer_values.cpp]: This literally could be a stand alone function :stuck_out_tongue:

File: src/goto-analyzer/get_function_pointer_values.cpp
Comment (thk123) [inline src/goto-analyzer/get_function_pointer_values.cpp]: :bulb: This function would be unit testable (and more generally usable) if you have it return the map, and have the caller write to the file

File: src/goto-analyzer/get_function_pointer_values.cpp
Comment (thk123) [inline src/goto-analyzer/get_function_pointer_values.cpp]: This is essentially a workaround right? {a, TOP, b} === TOP right? I'd pull this out into a function, and comment explaining why this is a problem

File: regression/goto-analyzer-get-fp-values/basic-usage-test/test.c
Comment (thk123) [inline regression/goto-analyzer-get-fp-values/basic-usage-test/test.c:62]: :bulb: It might make the examples easier to follow if they returned 'f' 'g' or 'h'

File: regression/goto-analyzer-get-fp-values/basic-usage-test/test.desc
Comment (thk123) [inline regression/goto-analyzer-get-fp-values/basic-usage-test/test.desc:24]: Am I correct in thinking these regexs check the json?

File: src/analyses/variable-sensitivity/abstract_enviroment.h
Comment (thk123) [inline src/analyses/variable-sensitivity/abstract_enviroment.h]: :question: (will check commit message) why does this need to become unprotected?

**Review by thk123 — COMMENTED** [review]
Reviewed: Get functions pointers into json...Update documentation

Looks mostly good - I have a few questions I'd like answered. Also could you confirm if there is a test that causes the code to generate duplicates in the set? Obviously storing them in the set discards them, but important that that behavior is covered.

File: src/analyses/variable-sensitivity/variable_sensitivity_domain.h
Comment (thk123) [reply src/analyses/variable-sensitivity/variable_sensitivity_domain.h:117]: :no_entry_sign: Commit message does not explain why - could you pull this into own commit and justify why the API is being expanded?

File: src/analyses/variable-sensitivity/abstract_enviroment.h
Comment (thk123) [reply src/analyses/variable-sensitivity/abstract_enviroment.h]: :no_entry_sign: Commit message does not explain why - could you pull this into own commit and justify why the API is being expanded?

**Review by thk123 — COMMENTED** [review]
Expanding the API deserves its own commit with justification of why necessary

File: src/analyses/variable-sensitivity/variable_sensitivity_domain.h
Comment (hannes-steffenhagen-diffblue) [reply src/analyses/variable-sensitivity/variable_sensitivity_domain.h:117]: We need the ability to evaluate individual expressions to abstract objects from outside the VSD itself to get at the abstract objects... this just isn't something we had to do before, which is why it wasn't public.

We can put this into a commit message once you're happy with the other changes.

File: regression/goto-analyzer-get-fp-values/basic-usage-test/test.desc
Comment (hannes-steffenhagen-diffblue) [reply regression/goto-analyzer-get-fp-values/basic-usage-test/test.desc:24]: yes

File: src/goto-analyzer/get_function_pointer_values.cpp
Comment (thk123) [inline src/goto-analyzer/get_function_pointer_values.cpp:133]: :bulb: Given we are in the context of `get_function_pointer_value` you might be able to drop the prefix, and just use `values_for_function` to reduce the boilerplate

File: src/goto-analyzer/get_function_pointer_values.cpp
Comment (thk123) [inline src/goto-analyzer/get_function_pointer_values.cpp:80]: :bulb: General feedback on this function: there is a lot of noise, which makes it hard to read the real work here. Some tips that might help:

 - drop `!= nullptr`; it is idiomatic in C++ to do `if(x)` for `if(x != nullptr)`
 - pull out the computation of `state_after_assignment` with all of the preconditions if they are really necessary
 - use empty lines to give the important lines space:
```C++
  INVARIANT(
    state_after_assignment_unique_ptr != nullptr,
    "we should always be able to query the abstract state after an assignment");
  auto const state_after_assignment =
    dynamic_cast<const variable_sensitivity_domaint *>(
      state_after_assignment_unique_ptr.get());
  INVARIANT(
    state_after_assignment != nullptr,
    "the context calling this is responsible for making sure the abstract"
    " interpretation happens in the variable sensitivity domain");

   auto const restriction_abstract_object =
    state_after_assignment->eval(restriction_symbol, args.ns);

  CHECK_RETURN(restriction_abstract_object != nullptr);
```

File: src/goto-analyzer/get_function_pointer_values.cpp
Comment (thk123) [inline src/goto-analyzer/get_function_pointer_values.cpp]: Looking at the docs and the implementation, I think this is guaranteed, so you could even drop it

File: src/goto-analyzer/get_function_pointer_values.cpp
Comment (thk123) [inline src/goto-analyzer/get_function_pointer_values.cpp:115]: :pick: declare this closer to usage

**Review by thk123 — COMMENTED** [review]
Thank you for taking the time to pull apart the central function - it is much easier to follow now. None of the below comments are blocking, but will wait for commit message tidy up before hitting approve, but lgtm :+1:

File: src/goto-analyzer/get_function_pointer_values.cpp
Comment (hannes-steffenhagen-diffblue) [reply src/goto-analyzer/get_function_pointer_values.cpp]: It is guaranteed. If it wasn't guaranteed it wouldn't be an invariant.

**Comment by thk123** [comment]
The CI failure seems to definitely be spurious - the same download link is used successfully by me, and by the develop build. I'm confused restarting it hasn't fixed it, and I'd be tempted to give it one more try - but I'm OK with this being merged without that job passing

**Review by thk123 — APPROVED** [review]
Commit history has been tidied up. In future, I think the first commit would be better as smaller chunks:

 - adding the infastrcture for new testing folder
 - adding the test
 - deleting some old unneeded code
 - the implementation itself
 - adding in the flag and associated plumbing

Since the changes are quite isolated, I guess it wouldn't be too hard to do that now anyway, if you were motivated, but not a blocking comment.

**Comment by thk123** [comment]
Code cov report looks troubling - looks like none of the tests are being executed on CI (at least not by code cov...)

File: regression/CMakeLists.txt
Comment (thk123) [inline regression/CMakeLists.txt:57]: This change needs to be reflected in the `regression/Makefile` I believe to get the tests to run on Make builds

Comment (NlightNFotis) [reply regression/CMakeLists.txt:57]: Done.

Weird this wasn't manifesting before. I just broke up the first commit, so there is no code changes - only different commits.

By the way, the github view for commit history is borked. The commits are alright but they show differently than in the command line git log.

**Pull request merged**
